### PR TITLE
Improve UAP Runner logging and enable showprogress flag

### PR DESCRIPTION
**Pull request by safern**
This will write the tests logs to a txt file that the WindowsAppLauncher is already reading at the end of the Application's execution and printing it out into the console. This is a step towards having a live output and also if we get a test crash we will have the output before the test crashed printed out to the console.

Also I enabled the -showprogress flag that can be passed to the runner as an argument and will print out when a test started and finished running. This helps for hangs and crashes.

cc: @tarekgh @danmosemsft @joperezr

### CODE REVIEW

File: src/xunit.runner.uap/Helpers.cs [hunk 1]
```diff
@@ -15,5 +16,14 @@ internal static class Helpers
             return await folder.CreateFileAsync(test, CreationCollisionOption.ReplaceExisting);
         }
 
+        public static StreamWriter GetStreamToFileInLocalStorage(string pathInLocalStorage)
+        {
+            string appSharedPath = Directory.GetParent(Windows.Storage.ApplicationData.Current.LocalFolder.Path).FullName;
```
Comment (tarekgh) [inline src/xunit.runner.uap/Helpers.cs]: >GetParent [](start = 45, length = 9)

why getting the parent?

File: src/xunit.runner.uap/StandardUapVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+namespace Xunit.ConsoleClient
+{
+    public class StandardUapVisitor : XmlTestExecutionVisitor
+    {
+        string assemblyName;
+        readonly ConcurrentDictionary<string, ExecutionSummary> completionMessages;
+        readonly StreamWriter log;
+        readonly bool showProgress;
+        readonly bool failSkips;
```
Comment (tarekgh) [inline src/xunit.runner.uap/StandardUapVisitor.cs]: could you follow the code notations we use? I mean put "private" specifier and would be good to name the fields prefixed with underscore so you don't have to use "this.xxx" while accessing it

File: src/xunit.runner.uap/StandardUapVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            log.WriteLine($"      {ExceptionUtility.CombineMessages(testFailed).Replace(Environment.NewLine, Environment.NewLine + "      ")}");
+
+            WriteStackTrace(ExceptionUtility.CombineStackTraces(testFailed));
+
+            return base.Visit(testFailed);
+        }
+
+        protected override bool Visit(ITestPassed testPassed)
+        {
+            return base.Visit(testPassed);
```
Comment (tarekgh) [inline src/xunit.runner.uap/StandardUapVisitor.cs]: >return base.Visit(testPassed); [](start = 12, length = 30)

do we need to count the failed and passed test to produce the summary at the end of the run?

File: src/xunit.runner.uap/StandardUapVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        void WriteStackTrace(string stackTrace)
+        {
+            if (String.IsNullOrWhiteSpace(stackTrace))
+                return;
+
+            log.WriteLine("      Stack Trace:");
+
+            foreach (var stackFrame in stackTrace.Split(new[] { Environment.NewLine }, StringSplitOptions.None))
+            {
+                log.WriteLine($"         {StackFrameTransformer.TransformFrame(stackFrame, Directory.GetCurrentDirectory())}");
```
Comment (tarekgh) [inline src/xunit.runner.uap/StandardUapVisitor.cs]: >GetCurrentDirectory [](start = 101, length = 19)

does the tests always switch to the test directory as current directory?

File: src/xunit.runner.uap/Helpers.cs [hunk 5]
```diff
@@ -15,5 +16,14 @@ internal static class Helpers
             return await folder.CreateFileAsync(test, CreationCollisionOption.ReplaceExisting);
```
Comment (tarekgh) [inline src/xunit.runner.uap/Helpers.cs]: we should change this one too if possible

File: src/xunit.runner.uap/StandardUapVisitor.cs (hunk 2 above)
Comment (safern) [reply src/xunit.runner.uap/StandardUapVisitor.cs]: Yes, forgot to update that :)

File: src/xunit.runner.uap/StandardUapVisitor.cs (hunk 4 above)
Comment (safern) [reply src/xunit.runner.uap/StandardUapVisitor.cs]: Yes, I followed the approach of what he have on our netcoreapp xunit runner.

File: src/xunit.runner.uap/StandardUapVisitor.cs (hunk 3 above)
Comment (safern) [reply src/xunit.runner.uap/StandardUapVisitor.cs]: Nope :), xunit will handle that for us, the base class will take care of it and then when test assembly is finished executing we will get the results on the Visit method.

https://github.com/dotnet/buildtools/pull/1667/files#diff-ee71c3a2de1958c609bbdc6fac71d0adR58

File: src/xunit.runner.uap/Helpers.cs (hunk 5 above)
Comment (safern) [reply src/xunit.runner.uap/Helpers.cs]: I will follow up with the testResults.xml in another PR, first the output :)

File: src/xunit.runner.uap/Helpers.cs (hunk 1 above)
Comment (safern) [reply src/xunit.runner.uap/Helpers.cs]: In order to access the SharedLocalFolder (the package root folder) through `ApplicationData` we need special permissions, so I'm getting the LocalFolder which will be something like: `C:\Users\safern\AppData\Local\Packages\5cd54353-3ed7-4a6e-a72f-db349f28867c_v52bfwc2c21ha\Local\` so I will get the parent which is the root folder and then create the file where the Application Launcher expects it. I could change it to be under the `LocalFolder` also.

File: src/xunit.runner.uap/MainPage.xaml.cs [hunk 6]
```diff
@@ -15,22 +14,40 @@ public sealed partial class MainPage : Page
...
             // Run tests for assemblies in current directory
             XunitTestRunner runner = new XunitTestRunner();
+            log = Helpers.GetStreamToFileInLocalStorage(@"AC\Temp\stdout.txt");
+
+            Task.Run(() => runner.RunTests(App.LaunchArgs.Arguments, log));
+        }
+
+        static void OnUnhandledException(object sender, UnhandledExceptionEventArgs e)
+        {
+            var ex = e.Exception as Exception;
```
Comment (danmosemsft) [inline src/xunit.runner.uap/MainPage.xaml.cs]: isn't it `e.ExceptionObject`

Comment (safern) [reply src/xunit.runner.uap/MainPage.xaml.cs]: For Windows runtime is `Exception` (https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.unhandledexceptioneventargs)

Comment (safern) [reply src/xunit.runner.uap/MainPage.xaml.cs]: My mistake. I think I'm using the wrong event handler. Will update :)

**Comment by safern** [comment]
@tarekgh I've addressed your PR to not use the Parent directory and instead use the LocalFolder.

**Comment by tarekgh** [comment]
LGTM

**Comment by mmitche** [comment]
@dotnet-bot test cross please (testing)

**Comment by mmitche** [comment]
@dotnet-bot test cross please (testing)

**Comment by safern** [comment]
Merging now that roslyn was updated in corefx and we can get an updated buildtools.

**Pull request merged**